src/main.rs

```diff
@@ -76,6 +68,49 @@ use smp::PerCpu;
 use sync::Once;
 use umode::UmodeTask;
 
+#[cfg(test)]
+mod tests {
+    use super::*;
+    pub trait Testable {
+        fn run(&self) -> TestResult;
+    }
+
+    impl<T> Testable for T
+    where
+        T: Fn() -> TestResult,
+    {
+        fn run(&self) -> TestResult {
+            println!("---{}---\t", core::any::type_name::<T>());
+            self()
+        }
+    }
+
+    pub fn test_runner(tests: &[&dyn Testable]) {
+        println!("Running {} tests\n", tests.len());
+        for test in tests {
+            match test.run() {
+                Ok(()) => println!("pass"),
+                Err(TestFailure::Fail) => println!("FAIL"),
+                Err(TestFailure::FailedAt(loc)) => println!("FAILED at {loc}"),
+            }
+        }
+    }
+
+    #[test_case]
+    fn example_test() -> TestResult {
+        Ok(())
+    }
+
+    #[test_case]
+    fn example_fail() -> TestResult {
+        Err(TestFailure::Fail)
+    }
+
+    #[test_case]
+    fn example_fail_at() -> TestResult {
+        Err(TestFailure::FailedAt("example_fail_at"))
+    }
```

Comment:
drop these? you can put examples in a doc comment

Reply:
fixed